README.md
# php-view-source-protection
protect a view source, since f12 and right click can be disabled, so for protecting viewsource this is the way (mandalorian) haha

# emergency release

if you using D5 version update it to 2.0.3.2 D5.1 E-F<br>
`fixing unwanted single filtration on firefox`<br>
2.0.4 D6r <b>release will be postponed until newyear</b> due to developers school exam<br>
fixing various D5 bug: like unwanted session and cookies<br>
this version is not suitable for production!<br>
i expect to be production ready when hits version 3 <br><br>

# quick how to use
├&nbsp;index.php (can be anything too)<br>
&nbsp;&nbsp;&nbsp;└ require "<your_OWN_directory>/viewsource_clean.php";<br>
  (can be anywhere in the world lol i mean folder)<br>
```
<?php
require /var/www/html/anywhere/anywhare/viewsource_clean.php;
/*codes here*/
```
<br>

# new version released! Ver 2.0.3 D5 , PHP 7.X
__introducing secure page builder (read the docs)<br>
*__sample websites ready to run__*<br>
https://github.com/zonblade/php-view-source-protection/tree/main/module/for%20non%20php%20programmers/simple%20websites<br><br>
*__Sample Codes__*<br><br>
https://retas.dev/demo_script/viewsource.php?about=load&samplecodes<br>
*__whats new?__*<br>
+ multi param! ?home=show&param2=x&param3=x&and_so_on<br>
+ safer method
+ build in redirect option
+ build in anti rightclick, ctrl+u & f12
<br>

# about this code
this is raw code, edit it as fits your system build with basic logic.<br>
this code was <b><u>intended to protect HTML files</u></b> using simple render tools.
which is usefull for UI/UX demo sites.<br>
<b> developer of this code (me), are still active developing this code, feel free to post an issue or request</b><br>
feel free to modify this code as you need :)

+ why do i even need this if its is cant hide it completely?
you know, view-source is impossible to hide! but now your dream come true! you can hide your viewsource ;)
+ for now this script can only do some trick , but i'm sure there's something else we can do with this logic in the future.

# docs
LANG_ID dokumentasi berbahasa indonesia<br>
https://github.com/zonblade/php-view-source-protection/wiki/(active)-ViewSource-Switch
<br><br>
LANG_EN documentation<br>
soon
<br><br>
live DEMO:<br>
https://retas.dev/demo_script/viewsource.php<br>
latest version (ready to use) check here :<br>
https://github.com/zonblade/php-view-source-protection/releases<br>

# tested browser
+ all modern browsers

# License
under MIT license
+ you can help me improve this code

## Created by 
obviously its me lol,<br>
instagram.com/zonblade

## changelog
+ parameter management (multi parameter);
+ *updated Documentation ID*
+ obliterate viewsource mode added
+ changing the code entirely
+ <b>ADD PARAMETER MANAGEMENT</B>
+ viewsource_protect.php (depreciated, but still exist, please use NOPARAM instead of this)
+ removing only LOAD parameter
+ Adding Load Render fixing Firefox bug.
+ Adding Firefox compability
+ Fixing Firefox compability


changelog.md
<h1>Changelog</h1>
Version Official Release 2.0.0 .
.
.

+ removing Class
+ <b>reason</b> : PHP behave very weirdly on firefox so i wont use class anymore.
+ adding new features using cookies!
+ changing function name to <b>vsp_start(),vsp_inspect()</b>
+ old version still available if you want to do experiment with it


viewsource_switch.php
<?php
function vsp_loader($dir)
{
    return define("VSP_LOADER",$dir);
};
function vsp_inspect($var)
{
    /*
        kode javascript didapatkan dari stackoverflow 
        https://stackoverflow.com/a/53511363
        thanks to PK-1825
        */
    $inspect = '
            <script type="text/javascript">
            document.addEventListener("contextmenu", event => event.preventDefault());
            window.onload = function () {
            document.addEventListener("contextmenu", function (e) {
            e.preventDefault();
            }, false);
            document.addEventListener("keydown", function (e) {
            //document.onkeydown = function(e) {
            // "I" key
            if (e.ctrlKey && e.shiftKey && e.keyCode == 73) {
            disabledEvent(e);
            }
            // "J" key
            if (e.ctrlKey && e.shiftKey && e.keyCode == 74) {
            disabledEvent(e);
            }
            // "S" key + macOS
            if (e.keyCode == 83 && (navigator.platform.match("Mac") ? e.metaKey : e.ctrlKey)) {
            disabledEvent(e);
            }
            // "U" key
            if (e.ctrlKey && e.keyCode == 85) {
            disabledEvent(e);
            }
            // "F12" key
            if (event.keyCode == 123) {
            disabledEvent(e);
            }
            }, false);
            function disabledEvent(e) {
            if (e.stopPropagation) {
            e.stopPropagation();
            } else if (window.event) {
            window.event.cancelBubble = true;
            }
            e.preventDefault();
            return false;
            }
            }
            </script>
            ';
    if($var == true){
        echo $inspect;
    }
};
function vsp_start($param,$dir,$load_dir)
{
    session_start();
    /*
    getting other params to save while loading
    */
    $uri_current = $_SERVER['REQUEST_URI'];
    $uri_confirm = substr($uri_current, strpos($uri_current, "&") + 1);
    if(strpos($uri_current, '&') !== false){
        $back_param = "&".$uri_confirm;
    }else{
        $back_param = '';
    }
    /*
    filter for not looping if got same param as main parameter!
    contohnya
    ?home=show&home=blabla > akan redirect ke ?home=load
    ?invalid=show&about= > akan redirect ke ?about=load
    */
    if(strpos($back_param, $param) !== false)
    {
        /*
        fungsi clear cookies & redirect
        */
        setcookie("vsp", "", time() - 3600);
        if($load_dir == true){include VSP_LOADER;};
        header("Refresh:0;url=?$param=load",true,301);
        die();
    }
    /*
    checking up if everything exist
    */
    if(isset($_GET) && !isset($_GET[$param]) && empty($_GET))
    {
        /*
        fungsi clear cookies & redirect
        */
        setcookie("vsp", "", time() - 3600);
        if($load_dir == true){include VSP_LOADER;};
        header("Refresh:0;url=?$param=load$back_param",true,301);
        die();
    }
    /*
    getting the real deal
    */
    switch($_GET[$param]){
        case "show":
            if(empty($_COOKIE['vsp']))
            {
                /*
                fungsi clear cookies & redirect
                */
                setcookie("vsp", "", time() - 3600);
                if($load_dir == true){include VSP_LOADER;};
                header("Refresh:0;url=?$param=load$back_param",true,301);
                die();
            }else{
                if($dir == false)
                {
                    setcookie("vsp", "", time() - 3600);
                }
                else
                {
                    setcookie("vsp", "", time() - 3600);
                    include $dir;
                    die();
                }
            }
            break;
        case 'load':
            /*
            fungsi load, disinilah kuncinya
            */
            setcookie("vsp", "", time() - 3600);
            setcookie("vsp", "true");
            if($load_dir == true){include VSP_LOADER;};
            header("Refresh:0;url=?$param=show$back_param",true,301);
            die();
        default:
            if(empty($_COOKIE['vsp']))
            {
                /*
                fungsi clear cookies & redirect
                */
                setcookie("vsp", "", time() - 3600);
                if($load_dir == true){include VSP_LOADER;};
                header("Refresh:0;url=?$param=load$back_param",true,301);
                die();
            }
            else
            {
                if($dir == false)
                {
                    setcookie("vsp", "", time() - 3600);
                }
                else
                {
                    setcookie("vsp", "", time() - 3600);
                    include $dir;
                    die();
                }
            }
    }
};
function vsp_init($param,$load_dir)
{
    if(empty($_COOKIE['vsp'])){
        /*
        fungsi clear cookies & redirect
        */
        setcookie("vsp", "", time() - 3600);
        if($load_dir == true){include VSP_LOADER;};
        header("Refresh:0;url=?$param=load",true,301);
        die();
    }
};
function vsp_builder($type,$param,$dir,$load_dir)
{
    if($type == 'load')
    {
        if(isset($_GET[$param]) && $dir != 'redirect')
        {
            vsp_start(  $param,$dir,$load_dir);
            vsp_init(   $param,$load_dir);
            die();
        }
    }

    if($type == 'redirect')
    {
        if(isset($_GET[$param]) && $load_dir == true)
        {
            header("location: $dir");
            die();
        }
        elseif(isset($_GET[$param]) && $load_dir == false)
        {
            header("location: ?");
            die();
        }
    }
};
function vsp_default($var,$dir,$redir_url)
{
    if($dir != 'redirect' && $redir_url == false)
    {
        if($var == true){
            include $dir;
        }
    }
    elseif($dir == 'redirect' && $redir_url != false)
    {
        header("location: ?$redir_url=");
        die();
    }
};
function vsp_kill()
{
    session_start();
    setcookie("vsp", "", time() - 3600);
};


viewsource_protect.php
<?php
function view_source_lock($var,$get_param,$get_value,$view_path,$load,$error){
    session_start();
    if($var == true){
        switch($_GET[$get_param]){
            /*
            ID : untuk case page load
            EN : to initialize load page
            */
            case $get_value:
                if($_SESSION["protect"] != true){
                    /*
                    ID : cek session protectnya
                    EN : checking the session
                    */
                    header("location: ?$get_param=$get_value-load");
                }else{
                    /*
                    ID : include halaman view/controler
                    EN : including the view/controller, as you wish
                    */
                    include $view_path;
                    /*
                    ID :
                    session protectnya cukup sampai sini,
                    agar jika reload akan kembali ke view load
                    ##
                    EN :
                    turning session to false,
                    this is mandatory to protect the view source.
                    */
                    $_SESSION["protect"] = false;
                    die();
                }
                break;
            case $get_value.'-load':
                /*
                ID :
                cek apakah halaman load diaktifkan?
                jika iya set load view dan set time ke 2 detik
                ##
                EN :
                checking if you have inputed the design for the load or not
                if its not it will set refresh value to 0 sec.
                */
                if($load != null){
                    include $load;
                    $time = 2; /* refresh after x sec */
                }else{
                    $time = 0;
                }
                /*
                ID :
                pada case ini session protect diaktifkan
                ##
                EN :
                acivating the session protect
                */
                $_SESSION["protect"] = true;
                header("Refresh:$time;url=?$get_param=$get_value");
                die();
                break;
            default:
                /*
                ID : menambahkan halaman error/404, ini penting
                EN : adding the error/notfound page, this is important
                */
                include $error;
        }
    }else{
        /*
        ID : kalau viewsource protectnya false langsung masuk halaman
        EN : if the protect view turn into false, it will directly gone into the page mentioned, get and param will be ignored.
        */
        include $view_path;
        die();
    }
}


code draft/unstable.php
<?php
function vsp_loader($dir)
{
    return define("VSP_LOADER",$dir);
};
function vsp_inspect($var)
{
    /*
        kode javascript didapatkan dari stackoverflow 
        https://stackoverflow.com/a/53511363
        thanks to PK-1825
        */
    $inspect = '
            <script type="text/javascript">
            document.addEventListener("contextmenu", event => event.preventDefault());
            window.onload = function () {
            document.addEventListener("contextmenu", function (e) {
            e.preventDefault();
            }, false);
            document.addEventListener("keydown", function (e) {
            //document.onkeydown = function(e) {
            // "I" key
            if (e.ctrlKey && e.shiftKey && e.keyCode == 73) {
            disabledEvent(e);
            }
            // "J" key
            if (e.ctrlKey && e.shiftKey && e.keyCode == 74) {
            disabledEvent(e);
            }
            // "S" key + macOS
            if (e.keyCode == 83 && (navigator.platform.match("Mac") ? e.metaKey : e.ctrlKey)) {
            disabledEvent(e);
            }
            // "U" key
            if (e.ctrlKey && e.keyCode == 85) {
            disabledEvent(e);
            }
            // "F12" key
            if (event.keyCode == 123) {
            disabledEvent(e);
            }
            }, false);
            function disabledEvent(e) {
            if (e.stopPropagation) {
            e.stopPropagation();
            } else if (window.event) {
            window.event.cancelBubble = true;
            }
            e.preventDefault();
            return false;
            }
            }
            </script>
            ';
    if($var == true){
        echo $inspect;
    }
};
function vsp_start($param,$dir,$load_dir)
{
    session_start();
    /*
    getting other params to save while loading
    */
    $uri_current = $_SERVER['REQUEST_URI'];
    $uri_confirm = substr($uri_current, strpos($uri_current, "&") + 1);
    if(strpos($uri_current, '&') !== false){
        $back_param = "&".$uri_confirm;
    }else{
        $back_param = '';
    }
    /*
    filter for not looping if got same param as main parameter!
    contohnya
    ?home=show&home=blabla > akan redirect ke ?home=load
    ?invalid=show&about= > akan redirect ke ?about=load
    */
    if(strpos($back_param, $param) !== false)
    {
        /*
        fungsi clear cookies & redirect
        */
        $_SESSION['vsp'] = '';
        setcookie("vsp", "", time() - 3600);
        if($load_dir == true){include VSP_LOADER;}else{echo '';};
        header("Refresh:0;url=?$param=load",true,301);
        die();
    }
    /*
    checking up if everything exist
    */
    if(isset($_GET) && !isset($_GET[$param]) && empty($_GET))
    {
        /*
        fungsi clear cookies & redirect
        */
        $_SESSION['vsp'] = '';
        setcookie("vsp", "", time() - 3600);
        if($load_dir == true){include VSP_LOADER;}else{echo '';};
        header("Refresh:0;url=?$param=load$back_param",true,301);
        die();
    }
    /*
    getting the real deal
    */
    switch($_GET[$param]){
        case "show":
            if(empty($_COOKIE['vsp']) && empty($_SESSION['vsp']))
            {
                /*
                fungsi clear cookies & redirect
                */
                $_SESSION['vsp'] = '';
                setcookie("vsp", "", time() - 3600);
                if($load_dir == true){include VSP_LOADER;}else{echo '';};
                header("Refresh:0;url=?$param=load$back_param",true,301);
                die();
            }else{
                if($dir == false)
                {
                    $_SESSION['vsp'] = '';
                    setcookie("vsp", "", time() - 3600);
                }
                else
                {
                    $_SESSION['vsp'] = '';
                    setcookie("vsp", "", time() - 3600);
                    include $dir;
                    die();
                }
            }
            break;
        case 'load':
            /*
            fungsi load, disinilah kuncinya
            */
            $_SESSION['vsp'] = '';
            setcookie("vsp", "", time() - 3600);
            $_SESSION['vsp'] = true;
            setcookie("vsp", "true");
            if($load_dir == true){include VSP_LOADER;}else{echo '';};
            header("Refresh:0;url=?$param=show$back_param",true,301);
            die();
        default:
            if(empty($_COOKIE['vsp']) && empty($_SESSION['vsp']))
            {
                /*
                fungsi clear cookies & redirect
                */
                $_SESSION['vsp'] = '';
                setcookie("vsp", "", time() - 3600);
                if($load_dir == true){include VSP_LOADER;}else{echo '';};
                header("Refresh:0;url=?$param=load$back_param",true,301);
                die();
            }
            else
            {
                if($dir == false)
                {
                    $_SESSION['vsp'] = '';
                    setcookie("vsp", "", time() - 3600);
                }
                else
                {
                    $_SESSION['vsp'] = '';
                    setcookie("vsp", "", time() - 3600);
                    include $dir;
                    die();
                }
            }
    }
};
function vsp_init($param,$load_dir)
{
    if(empty($_COOKIE['vsp'])){
        /*
        fungsi clear cookies & redirect
        */
        session_start();
        $_SESSION['vsp'] = '';
        setcookie("vsp", "", time() - 3600);
        if($load_dir == true){include VSP_LOADER;}else{echo '';};
        header("Refresh:0;url=?$param=load",true,301);
        die();
    }
};
function vsp_builder($type,$param,$dir,$load_dir)
{
    if($type == 'load')
    {
        if(isset($_GET[$param]) && $dir != 'redirect')
        {
            vsp_start(  $param,$dir,$load_dir);
            vsp_init(   $param,$load_dir);
            die();
        }
    }

    if($type == 'redirect')
    {
        if(isset($_GET[$param]) && $load_dir == true)
        {
            header("location: $dir");
            die();
        }
        elseif(isset($_GET[$param]) && $load_dir == false)
        {
            header("location: ?");
            die();
        }
    }
};
function vsp_default($var,$dir,$redir_url)
{
    if($dir != 'redirect' && $redir_url == false)
    {
        if($var == true){
            include $dir;
        }
    }
    elseif($dir == 'redirect' && $redir_url != false)
    {
        header("location: ?$redir_url=");
        die();
    }
};
function vsp_kill()
{
    session_start();
    $_SESSION['vsp'] = '';
    setcookie("vsp", "", time() - 3600);
};


depreciated/viewsource.php
<?php
class vsp {
    /* 
    penjelasannya mirip ya seperti viewsource_protect
    untuk memahaminya silahkan melihat file yang satu lagi, penjelasannya jelas disana.
    */
    function loader($dir,$time){
        if($dir != null){
            $this->loader       = $dir;
            $this->loader_time  = $time;
        }else{
            $this->loader_time  = 0;
        }
    }
    function inspect($var){
        /*
        kode javascript didapatkan dari stackoverflow 
        https://stackoverflow.com/a/53511363
        thanks to PK-1825
        */
        $this->inspect = '
            <script type="text/javascript">
            document.addEventListener("contextmenu", event => event.preventDefault());
            window.onload = function () {
            document.addEventListener("contextmenu", function (e) {
            e.preventDefault();
            }, false);
            document.addEventListener("keydown", function (e) {
            //document.onkeydown = function(e) {
            // "I" key
            if (e.ctrlKey && e.shiftKey && e.keyCode == 73) {
            disabledEvent(e);
            }
            // "J" key
            if (e.ctrlKey && e.shiftKey && e.keyCode == 74) {
            disabledEvent(e);
            }
            // "S" key + macOS
            if (e.keyCode == 83 && (navigator.platform.match("Mac") ? e.metaKey : e.ctrlKey)) {
            disabledEvent(e);
            }
            // "U" key
            if (e.ctrlKey && e.keyCode == 85) {
            disabledEvent(e);
            }
            // "F12" key
            if (event.keyCode == 123) {
            disabledEvent(e);
            }
            }, false);
            function disabledEvent(e) {
            if (e.stopPropagation) {
            e.stopPropagation();
            } else if (window.event) {
            window.event.cancelBubble = true;
            }
            e.preventDefault();
            return false;
            }
            }
            </script>
            ';
        if($var == true){
            echo $this->inspect;
        }
    }
    function render_opt($render,$path,$key){
        $this->render = $render;
        $this->path = $path;
        $this->keys = $key;
    }
    function render_load(){
        if(isset($_POST['flock_load'])){
            if($_POST['flock_load'] == $this->keys){
                include $this->path;
                die();
            };
        };
    }
    function start(){
        session_start();
        if($_GET['load'] == 'on'){
            if(isset($this->loader)){
                include $this->loader;
                $time = $this->loader_time;
            }else{
                $time = 0;
            }
            $_SESSION["protect"] = true;
            $link_situs = (isset($_SERVER['HTTPS']) && $_SERVER['HTTPS'] === 'on' ? "https" : "http") . "://$_SERVER[HTTP_HOST]$_SERVER[REQUEST_URI]";
            $link_situs = preg_replace('/(&|\?)'.preg_quote('load').'=[^&]*$/', '', $link_situs);
            $link_situs = preg_replace('/(&|\?)'.preg_quote('load').'=[^&]*&/', '$1', $link_situs);
            header("Refresh:$time;url=$link_situs");
            die();
        }else{
            switch($_SESSION['protect']){
                case false:
                    header("location: ?load=on");
                    die();
                    break;
                default :
                    $_SESSION['protect'] = false;
                    if($this->render == true){
                        echo '
<vsprender>';
include $this->loader;
echo '<!-- this is render stage --></vsprender>
<script id="load_1" src="https://code.jquery.com/jquery-3.5.1.min.js" integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossorigin="anonymous"></script>
<script id="load_2"type="text/javascript">
$.ajax({url:"?",method:"POST",data:{flock_load:"'.$this->keys.'"},success:function(o){$("html").html(o),$("#load_1").remove(),$("#load_2").remove()}});
</script>
';
                        die();
                    }else{
                        include $this->path;
                        die();
                    }
            }
        }
    }
    function stop(){
        session_start();
        $_SESSION['protect'] = false;
    }
    function param_load($strict_param,$strict_value,$file_dir){
        $this->StrictCase   = $strict_param;
        $this->StrictDir    = $file_dir;
        $this->StrictVal    = $strict_value;
    }
    function param_start(){
        switch($_GET[$this->StrictCase]){
            case "$this->StrictVal":
                session_start();
                if($_SESSION['strict'] != true){
                    header("location: ?$this->StrictCase=load");
                    die();
                }else{
                    require_once $this->StrictDir;
                    $_SESSION['strict'] = false;
                    die();
                }
                break;
            default:
                session_start();
                $_SESSION['strict'] = true;
                include $this->loader;
                header("Refresh:$this->loader_time; url=?$this->StrictCase=$this->StrictVal",true,301);
                die();
        }
    }
}


module/for non php programmers/simple websites/tools/viewsource_clean.php
<?php
function vsp_loader($dir)
{
    return define("VSP_LOADER",$dir);
};
function vsp_inspect($var)
{
    /*
        kode javascript didapatkan dari stackoverflow 
        https://stackoverflow.com/a/53511363
        thanks to PK-1825
        */
    $inspect = '
            <script type="text/javascript">
            document.addEventListener("contextmenu", event => event.preventDefault());
            window.onload = function () {
            document.addEventListener("contextmenu", function (e) {
            e.preventDefault();
            }, false);
            document.addEventListener("keydown", function (e) {
            //document.onkeydown = function(e) {
            // "I" key
            if (e.ctrlKey && e.shiftKey && e.keyCode == 73) {
            disabledEvent(e);
            }
            // "J" key
            if (e.ctrlKey && e.shiftKey && e.keyCode == 74) {
            disabledEvent(e);
            }
            // "S" key + macOS
            if (e.keyCode == 83 && (navigator.platform.match("Mac") ? e.metaKey : e.ctrlKey)) {
            disabledEvent(e);
            }
            // "U" key
            if (e.ctrlKey && e.keyCode == 85) {
            disabledEvent(e);
            }
            // "F12" key
            if (event.keyCode == 123) {
            disabledEvent(e);
            }
            }, false);
            function disabledEvent(e) {
            if (e.stopPropagation) {
            e.stopPropagation();
            } else if (window.event) {
            window.event.cancelBubble = true;
            }
            e.preventDefault();
            return false;
            }
            }
            </script>
            ';
    if($var == true){
        echo $inspect;
    }
};
function vsp_start($param,$dir,$load_dir)
{
    session_start();
    /*
    getting other params to save while loading
    */
    $uri_current = $_SERVER['REQUEST_URI'];
    $uri_confirm = substr($uri_current, strpos($uri_current, "&") + 1);
    if(strpos($uri_current, '&') !== false){
        $back_param = "&".$uri_confirm;
    }else{
        $back_param = '';
    }
    /*
    filter for not looping if got same param as main parameter!
    contohnya
    ?home=show&home=blabla > akan redirect ke ?home=load
    ?invalid=show&about= > akan redirect ke ?about=load
    */
    if(strpos($back_param, $param) !== false)
    {
        /*
        fungsi clear cookies & redirect
        */
        setcookie("vsp", "", time() - 3600);
        if($load_dir == true){include VSP_LOADER;}else{echo '';};
        header("Refresh:0;url=?$param=load",true,301);
        die();
    }
    /*
    checking up if everything exist
    */
    if(isset($_GET) && !isset($_GET[$param]) && empty($_GET))
    {
        /*
        fungsi clear cookies & redirect
        */
        setcookie("vsp", "", time() - 3600);
        if($load_dir == true){include VSP_LOADER;}else{echo '';};
        header("Refresh:0;url=?$param=load$back_param",true,301);
        die();
    }
    /*
    getting the real deal
    */
    switch($_GET[$param]){
        case "show":
            if(empty($_COOKIE['vsp']))
            {
                /*
                fungsi clear cookies & redirect
                */
                setcookie("vsp", "", time() - 3600);
                if($load_dir == true){include VSP_LOADER;}else{echo '';};
                header("Refresh:0;url=?$param=load$back_param",true,301);
                die();
            }else{
                if($dir == false)
                {
                    setcookie("vsp", "", time() - 3600);
                }
                else
                {
                    setcookie("vsp", "", time() - 3600);
                    include $dir;
                    die();
                }
            }
            break;
        case 'load':
            /*
            fungsi load, disinilah kuncinya
            */
            setcookie("vsp", "", time() - 3600);
            setcookie("vsp", "true");
            if($load_dir == true){include VSP_LOADER;}else{echo '';};
            header("Refresh:0;url=?$param=show$back_param",true,301);
            die();
        default:
            if(empty($_COOKIE['vsp']))
            {
                /*
                fungsi clear cookies & redirect
                */
                setcookie("vsp", "", time() - 3600);
                if($load_dir == true){include VSP_LOADER;}else{echo '';};
                header("Refresh:0;url=?$param=load$back_param",true,301);
                die();
            }
            else
            {
                if($dir == false)
                {
                    setcookie("vsp", "", time() - 3600);
                }
                else
                {
                    setcookie("vsp", "", time() - 3600);
                    include $dir;
                    die();
                }
            }
    }
};
function vsp_init($param,$load_dir)
{
    if(empty($_COOKIE['vsp'])){
        /*
        fungsi clear cookies & redirect
        */
        setcookie("vsp", "", time() - 3600);
        if($load_dir == true){include VSP_LOADER;}else{echo '';};
        header("Refresh:0;url=?$param=load",true,301);
        die();
    }
};
function vsp_builder($type,$param,$dir,$load_dir)
{
    if($type == 'load')
    {
        if(isset($_GET[$param]) && $dir != 'redirect')
        {
            vsp_start(  $param,$dir,$load_dir);
            vsp_init(   $param,$load_dir);
            die();
        }
    }

    if($type == 'redirect')
    {
        if(isset($_GET[$param]) && $load_dir == true)
        {
            header("location: $dir");
            die();
        }
        elseif(isset($_GET[$param]) && $load_dir == false)
        {
            header("location: ?");
            die();
        }
    }
};
function vsp_default($var,$dir,$redir_url)
{
    if($dir != 'redirect' && $redir_url == false)
    {
        if($var == true){
            include $dir;
        }
    }
    elseif($dir == 'redirect' && $redir_url != false)
    {
        header("location: ?$redir_url=");
        die();
    }
};
function vsp_kill()
{
    session_start();
    setcookie("vsp", "", time() - 3600);
};


viewsource_noparam.php
<?php
class vsp {
    /* 
    penjelasannya mirip ya seperti viewsource_protect
    untuk memahaminya silahkan melihat file yang satu lagi, penjelasannya jelas disana.
    */
    function loader($dir,$time){
        $this->loader       = $dir;
        $this->loader_time  = $time;
    }

    function render_opt($render,$path,$key){
        $this->render = $render;
        $this->path = $path;
        $this->keys = $key;
    }

    function render_load(){
        if(isset($_POST['flock_load'])){
            if($_POST['flock_load'] == $this->keys){
                include $this->path;
                die();
            };
        };
    }

    function inspect($var){
        /*
        kode javascript didapatkan dari stackoverflow 
        https://stackoverflow.com/a/53511363
        thanks to PK-1825
        */
        $this->inspect = '
            <script type="text/javascript">
            document.addEventListener("contextmenu", event => event.preventDefault());
            window.onload = function () {
            document.addEventListener("contextmenu", function (e) {
            e.preventDefault();
            }, false);
            document.addEventListener("keydown", function (e) {
            //document.onkeydown = function(e) {
            // "I" key
            if (e.ctrlKey && e.shiftKey && e.keyCode == 73) {
            disabledEvent(e);
            }
            // "J" key
            if (e.ctrlKey && e.shiftKey && e.keyCode == 74) {
            disabledEvent(e);
            }
            // "S" key + macOS
            if (e.keyCode == 83 && (navigator.platform.match("Mac") ? e.metaKey : e.ctrlKey)) {
            disabledEvent(e);
            }
            // "U" key
            if (e.ctrlKey && e.keyCode == 85) {
            disabledEvent(e);
            }
            // "F12" key
            if (event.keyCode == 123) {
            disabledEvent(e);
            }
            }, false);
            function disabledEvent(e) {
            if (e.stopPropagation) {
            e.stopPropagation();
            } else if (window.event) {
            window.event.cancelBubble = true;
            }
            e.preventDefault();
            return false;
            }
            }
            </script>
            ';
        if($var == true){
            echo $this->inspect;
        }
    }

    function start(){
        session_start();
        if($_GET['load'] == 'on'){
            if(isset($this->loader)){
                include $this->loader;
                $time = $this->loader_time;
            }else{
                $time = 0;
            }
            $_SESSION["protect"] = true;
            $link_situs = (isset($_SERVER['HTTPS']) && $_SERVER['HTTPS'] === 'on' ? "https" : "http") . "://$_SERVER[HTTP_HOST]$_SERVER[REQUEST_URI]";
            $link_situs = preg_replace('/(&|\?)'.preg_quote('load').'=[^&]*$/', '', $link_situs);
            $link_situs = preg_replace('/(&|\?)'.preg_quote('load').'=[^&]*&/', '$1', $link_situs);
            header("Refresh:$time;url=$link_situs");
            die();
        }else{
            switch($_SESSION['protect']){
                case false:
                    header("location: ?load=on");
                    die();
                    break;
                default :
                    $_SESSION['protect'] = false;
                    if($this->render == true){
                        echo '
<vsprender>
';
include $this->loader;
echo '
<!-- this is render stage -->
</vsprender>
<script id="load_1" src="https://code.jquery.com/jquery-3.5.1.min.js" integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossorigin="anonymous"></script>
<script id="load_2"type="text/javascript">
$.ajax({
url : "?",
method : "POST",
data: {
    "flock_load": "'.$this->keys.'",
},
success: function(data){
    $("html").html(data);
    $("#load_1").remove();
    $("#load_2").remove();
}
});
</script>
';
                        die();
                    }else{
                        include $this->path;
                    }
            }
        }
    }

    function stop(){
        session_start();
        $_SESSION['protect'] = false;
    }

    function start_render($render,$path,$key){
    }
}
